File: cs336_basics/bpe_tokenization/token_collection.py
import logging

from typing import TypeAlias
from collections import defaultdict, OrderedDict

TokenIdPair: TypeAlias = tuple[int, int]

logger = logging.getLogger()
# logging.basicConfig(level=logging.DEBUG)

# def info(type, value, tb):
#     if hasattr(sys, 'ps1') or not sys.stderr.isatty() or type != AssertionError:
#         # we are in interactive mode or we don't have a tty-like
#         # device, so we call the default hook
#         sys.__excepthook__(type, value, tb)
#     else:
#         import traceback, pdb
#         # we are NOT in interactive mode, print the exception...
#         traceback.print_exception(type, value, tb)
#         print
#         # ...then start the debugger in post-mortem mode.
#         pdb.pm()

# sys.excepthook = info

class TokenCollection:
    def __init__(self, value: bytes):
        self.token_ids = [b for b in value]
        self.next = [i for i in range(1, len(self.token_ids))] + [None]
        self.prev = [None] + [i for i in range(len(self.token_ids) - 1)]
        self._pair_first_idx: dict[TokenIdPair, OrderedDict[int, None]] = defaultdict(
            OrderedDict[int, None]
        )

        for i in range(len(self.token_ids) - 1):
            self._pair_first_idx[(self.token_ids[i], self.token_ids[i + 1])][i] = None

    @property
    def pairs_cnt(self) -> dict[TokenIdPair, int]:
        pairs_cnt: dict[TokenIdPair, int] = defaultdict(int)
        for idx in range(len(self.token_ids) - 1):
            pair = self.token_ids[idx], self.token_ids[idx + 1]
            pairs_cnt[pair] += 1
        return pairs_cnt

    def _replace_pair(self, first_idx: int, new_token_id: int):
        second_idx = self.next[first_idx]
        assert second_idx is not None

        self.token_ids[first_idx] = new_token_id
        new_next = self.next[second_idx]
        self.next[first_idx] = self.next[second_idx]
        if new_next:
            self.prev[new_next] = first_idx

        self.next[second_idx] = None
        self.prev[second_idx] = None
        self.token_ids[second_idx] = -1  # mark as removed

    def merge_bytes_pair(
        self, target_pair: TokenIdPair, new_token_id: int
    ) -> dict[TokenIdPair, int]:
        if (
            target_pair not in self._pair_first_idx
            or len(self._pair_first_idx[target_pair]) == 0
        ):
            return {}

        last_match_idx = None
        idx_to_merge: list[int] = []
        for idx in self._pair_first_idx[target_pair]:
            if self.prev[idx] is not None and self.prev[idx] == last_match_idx:
                continue
            last_match_idx = idx
            idx_to_merge.append(idx)

        # logger.debug("found %d occurrences", len(idx_to_merge))
        diff: dict[TokenIdPair, int] = defaultdict(int)
        t1, t2 = target_pair
        for merge_idx in idx_to_merge:
            t2_idx = self.next[merge_idx]
            assert t2_idx is not None, "Next index is None, cannot merge pair"
            # logger.debug("Merging tokens at idx: %d and idx: %d", t1_idx, t2_idx)
            # logger.debug("prev: %s, next: %s", self.prev[t1_idx], self.next[t2_idx])

            self._replace_pair(merge_idx, new_token_id)

            # look backward
            if (prev_idx := self.prev[merge_idx]) is not None:
                new_pair = (self.token_ids[prev_idx], new_token_id)
                diff[new_pair] += 1
                self._pair_first_idx[new_pair][prev_idx] = None

                old_pair = (self.token_ids[prev_idx], t1)
                diff[old_pair] -= 1
                del self._pair_first_idx[old_pair][prev_idx]

            # look forward
            if (next_idx := self.next[merge_idx]) is not None:
                new_pair = (new_token_id, self.token_ids[next_idx])
                diff[new_pair] += 1
                self._pair_first_idx[new_pair][merge_idx] = None

                old_pair = (t2, self.token_ids[next_idx])
                diff[old_pair] -= 1
                del self._pair_first_idx[old_pair][t2_idx]

        del self._pair_first_idx[target_pair]
        return diff

File: cs336_basics/bpe_tokenization/cc/token_collection.cc
#include "token_collection.h"
#include <iostream>

namespace bpe {

TokenCollection::TokenCollection(std::string_view text) {
  size_t text_length = text.length();
  assert(text_length <= std::numeric_limits<Index>::max());

  token_ids_.reserve(text_length);
  prev_.reserve(text_length);
  next_.reserve(text_length);

  for (Index i = 0; i < text_length ; ++i) {
    token_ids_.push_back(reinterpret_cast<const unsigned char&>(text[i]));
    prev_.push_back(i - 1);
    next_.push_back(i + 1);
  }
  next_[text_length - 1] = -1;

  for (Index i = 0; i < text_length - 1; ++i) {
    TokenIdPair pair = std::make_pair(token_ids_[i], token_ids_[i + 1]);
    pair_first_idx[pair].insert(i);
  }
}

TokenCollection::PairFreqMap TokenCollection::GetPairFreq() const {
  PairFreqMap pair_freq;
  for (const auto &[pair, indices] : pair_first_idx) {
    pair_freq[pair] = indices.size();
  }
  return pair_freq;
}

TokenCollection::PairFreqMap
TokenCollection::MergePair(const TokenIdPair& target_pair, TokenId new_token_id) {
  auto it = pair_first_idx.find(target_pair);
  if (it == pair_first_idx.end() || it->second.empty()) {
    return {};
  }

  PairFreqMap diff;
  const auto& [t1, t2] = target_pair;
  Index last_merge_idx = -1;
  std::vector<Index> idx_to_merge;
  for (const Index idx : it->second) {
    if (prev_[idx] >= 0 && prev_[idx] == last_merge_idx) {
      continue;
    }
    last_merge_idx = idx;
    idx_to_merge.push_back(idx);
  }

  for(Index merge_idx: idx_to_merge) {
    Index t2_idx = next_[merge_idx];
    assert(t2_idx != -1);

    ReplacePair(merge_idx, new_token_id);

    if (Index prev_idx = prev_[merge_idx]; prev_idx != -1) {
      TokenIdPair new_pair = std::make_pair(token_ids_[prev_idx], new_token_id);
      diff[new_pair] += 1;
      pair_first_idx[new_pair].insert(prev_idx);

      TokenIdPair old_pair = std::make_pair(token_ids_[prev_idx], t1);
      diff[old_pair] -= 1;
      pair_first_idx[old_pair].erase(prev_idx);
    }

    if (Index next_idx = next_[merge_idx]; next_idx != -1) {
      TokenIdPair new_pair = std::make_pair(new_token_id, token_ids_[next_idx]);
      diff[new_pair] += 1;
      pair_first_idx[new_pair].insert(merge_idx);

      TokenIdPair old_pair = std::make_pair(t2, token_ids_[next_idx]);
      diff[old_pair] -= 1;
      pair_first_idx[old_pair].erase(t2_idx);
    }
  }

  pair_first_idx.erase(target_pair);
  return diff;
}

void TokenCollection::ReplacePair(Index first_idx, TokenId new_token_id) {
  Index second_idx = next_[first_idx];
  assert(second_idx != -1);

  token_ids_[first_idx] = new_token_id;
  Index new_next = next_[second_idx];
  next_[first_idx] = new_next;
  if (new_next != -1) {
    prev_[new_next] = first_idx;
  }

  // Unnecessary as these will never be accessed again.
  // next[second_idx] = -1;
  // prev[second_idx] = -1;
  // token_ids[second_idx] = -1;
}

} // namespace bpe
